crates/storage/provider/src/providers/static_file/manager.rs

```diff
@@ -448,6 +452,161 @@ impl StaticFileProvider {
         Ok(())
     }
 
+    /// Ensures that any broken invariants return a pipeline target to unwind to.
+    ///
+    /// Two types of consistency checks are done for:
+    ///
+    /// 1) When a static file fails to commit but the underlying data was changed.
+    /// 2) When a static file was committed, but the required database transaction was not.
+    ///
+    /// For 1) it can self-heal. Opening a writer to this segment will automatically do that.
+    /// For 2) the invariants below are checked, and if broken, require a pipeline unwind to heal.
+    ///
+    /// For each static file segment:
+    /// * the corresponding database table should overlap or have continuity in their keys
+    ///   ([TxNumber] or [BlockNumber]).
+    /// * its highest block should match the stage checkpoint block number if it's equal or higher
+    ///   than the corresponding database table last entry.
+    ///
+    /// Returns a [`Option`] of [`PipelineTarget::Unwind`] if any healing is required.
+    pub fn check_consistency<TX: DbTx>(
+        &self,
+        provider: &DatabaseProvider<TX>,
+        has_receipt_pruning: bool,
+    ) -> ProviderResult<Option<PipelineTarget>> {
+        let mut unwind_target: Option<BlockNumber> = None;
+        let mut update_unwind_target = |new_target: Option<BlockNumber>| {
+            let new_target = new_target.unwrap_or_default();
+            if let Some(target) = unwind_target.as_mut() {
+                *target = (*target).min(new_target);
+            } else {
+                unwind_target = Some(new_target);
+            }
+        };
+
+        for segment in StaticFileSegment::iter() {
+            if has_receipt_pruning && segment.is_receipts() {
+                // Pruned nodes (including full node) do not store receipts as static files.
+                continue
+            }
+
+            let initial_highest_block = self.get_highest_static_file_block(segment);
+
+            // By calling a writer, an inner consistency check will be triggered on
+            // [`NippyJarWriter::check_consistency_and_heal`]
+            //
+            // * If appending data was interrupted before a config commit, then data file will be
+            //   truncated according to the config.
+            //
+            // * If pruning data was interrupted before a config commit, then we have deleted data
+            //   that we are expected to still have. We need to check the Database and unwind
+            //   everything accordingly.
+            let mut writer = self.latest_writer(segment)?;
+
+            // This will commit any inconsistencies found and update the reader index. If it
+            // happens, it means that the highest block of the segement has decreased.
+            writer.commit()?;
```

Comment:
can we call the `check_consistency_and_heal` directly, so it's not hidden behind the `self.latest_writer()` implementation that if changed in the future will break this code?